ptvsd/_util.py

```diff
@@ -0,0 +1,245 @@
+        """Add a resource to be closed when closing."""
+        close = resource.close
+        if before:
+            def handle_closing(before):
+                if not before:
+                    return
+                close()
+        else:
+            def handle_closing(before):
+                if before:
+                    return
+                close()
+        self.add_close_handler(handle_closing)
+
+    def add_close_handler(self, handle_closing, nodupe=True):
+        """Add a func to be called when closing.
+
+        The func takes one arg: True if it was called before the main
+        close func and False if after.
+        """
+        with self._closedlock:
+            if self._closed:
+                if self.FAIL_ON_ALREADY_CLOSED:
+                    raise ClosedError('already closed')
+                return
+            if nodupe and handle_closing in self._handlers:
+                raise ValueError('close func already added')
+
+            self._handlers.append(handle_closing)
+
+    def check_closed(self):
+        """Raise ClosedError if closed."""
+        if self._closed:
+            if self.NAME:
+                raise ClosedError('{} closed'.format(self.NAME))
+            else:
+                raise ClosedError('closed')
+
+    @contextlib.contextmanager
+    def while_not_closed(self):
+        """A context manager under which the object will not be closed."""
+        with self._closedlock:
+            self.check_closed()
+            yield
+
+    def close(self):
+        """Release any owned resources and clean up."""
+        with self._closedlock:
+            if self._closed:
+                if self.FAIL_ON_ALREADY_CLOSED:
+                    raise ClosedError('already closed')
+                return
+            self._closed = True
+            handlers = list(self._handlers)
+
+        results = call_all(handlers, True)
+        self._log_results(results)
+        self._close()
+        results = call_all(handlers, False)
+        self._log_results(results)
+
+    # implemented by subclasses
+
+    def _close(self):
+        pass
+
+    # internal methods
+
+    def _log_results(self, results, log=None):
+        if log is None:
+            return
+        for obj, exc in results:
+            if exc is None:
+                continue
+            log('failed to close {!r} ({!r})'.format(obj, exc))
+
+
+########################
+# running stuff
+
+class NotRunningError(RuntimeError):
+    """Something isn't currently running."""
+
+
+class AlreadyStartedError(RuntimeError):
+    """Something was already started."""
+
+
+class AlreadyRunningError(AlreadyStartedError):
+    """Something is already running."""
+
+
+class Startable(object):
+    """A base class for types that may be started."""
+
+    RESTARTABLE = False
+    FAIL_ON_ALREADY_STOPPED = True
+
+    def __init__(self):
+        super(Startable, self).__init__()
+        self._is_running = None
```

Comment:
Looks like `lambda: False` here would simplify the code that invokes this.

Reply:
fixed